### ISSUE TITLE
[2018 ICASSP] Generalized End-to-End Loss for Speaker Verification

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#98]
Speaker verification task에서 성능을 끌어올린 논문

논문에서 제안한 generalized end-to-end (GE2E) loss만 설명

### 1.1. Overview
<img width="800" alt="image" src="https://user-images.githubusercontent.com/49643709/201803613-325c1f6b-a0ac-435d-8f65-d1181bc49fa2.png">
<img width="400" alt="image" src="https://user-images.githubusercontent.com/49643709/201803644-08cf1985-da93-4fd4-aee7-5c1638814f19.png">

모든 batch의 utterance와 speaker centroid간의 similarity를 측정하여 같은 speaker는 가까워지게, 다른 speaker는 멀어지게 학습

### 1.2. Details

Scenario를 하나 가정하고 설명을 하는 것이 이해에 도움이 되기에, speaker A에 대한 index 0의 utterance에 대해 어떻게 loss가 구해지는지 서술

#### 1.2.1. Embedding vector

<img width="400" alt="image" src="https://user-images.githubusercontent.com/49643709/201808895-a7fabe21-50b6-4991-91ca-ef2a4830ac45.png">

Network output에 L2 normalization을 통해 embedding vector를 구함

#### 1.2.2. Centroids
<img width="400" alt="image" src="https://user-images.githubusercontent.com/49643709/201808957-448110dd-22ce-4eb6-b4a6-9d9b96c301af.png">
<img width="400" alt="image" src="https://user-images.githubusercontent.com/49643709/201809010-07f9295a-9572-4912-8d96-1a8e485177bb.png">

Speaker A의 centroid : index 0번을 제외한 batch 내에서의 speaker A embeddings의 mean을 통해 구함

Idx 0 data를 speaker A centroid에서 제외하는 이유 : makes training stable & helps avoid trivial solutions

Speaker A 이외의 다른 speaker의 centroid : batch 내에서의 각 speaker별 embeddings의 mean

#### 1.2.3. Similarity
<img width="400" alt="image" src="https://user-images.githubusercontent.com/49643709/201809116-dd44a48a-102c-43bb-bff9-d0a6860b8810.png">

Scaled cosine similarities

#### 1.2.4. Implementation
<img width="400" alt="image" src="https://user-images.githubusercontent.com/49643709/201809340-37d7ae63-07f0-4238-8f92-c539142e488f.png">
<img width="400" alt="image" src="https://user-images.githubusercontent.com/49643709/201810114-08ef0676-e03b-4819-84c1-bce90cb55471.png">

Positive centroid는 가깝게, negative centroid는 멀어지게 하는 concept을 implement하는 2가지 방식이 있음

1. Softmax : positive centroid는 가깝게, 나머지 모든 negative centroids는 멀어지게 학습
2. Contrast : positive centroid는 가깝게, hard negative centroid는 멀어지게 학습 

#### 1.2.5. Loss
<img width="400" alt="image" src="https://user-images.githubusercontent.com/49643709/201810417-7f8f4db4-5f27-4f8e-9dd0-1c9b2925d2d1.png">

각 data마다의 loss를 sum하여 사용

### 1.3. Pseudo Code

Pytorch style pseudo code of GE2E loss with softmax

```
Input dimension : (N, M, D)

N : number of speakers in a batch
M : number of utterances for each speaker
D : dimension of the speaker embeddings
```

``` python
class GE2E_loss(nn.Module):
    def __init__(self, init_w=10.0, init_b=-5.0, **kwargs):
        super(GE2E_loss, self).__init__()
        self.w = nn.Parameter(torch.tensor(init_w))
        self.b = nn.Parameter(torch.tensor(init_b))
        self.criterion  = nn.CrossEntropyLoss()

    def forward(self, x):
        N, M = x.shape[0], x.shape[1]
        centroids = torch.mean(x, 1)  # (N, D)
        centroids = centroids.unsqueeze(-1).transpose(0, 2)  # (1, D, N)

        cos_sim_matrix = []
        for i in range(M): 
            idx = [*range(M)]
            idx.remove(i)
            except_centroids = torch.mean(x[:,idx,:], 1)  # (N, D)
            cos_sim_diag = F.cosine_similarity(x[:,i,:], except_centroids)    # (N, )
            cos_sim = F.cosine_similarity(x[:,i,:].unsqueeze(-1), centroids)  # (N, N)
            cos_sim[range(N), range(N)] = cos_sim_diag
            cos_sim_matrix.append(torch.clamp(cos_sim, 1e-6))

        cos_sim_matrix = torch.stack(cos_sim_matrix, dim=1)  # (N, M, N)
        cos_sim_matrix = cos_sim_matrix.view(-1, N)  # (N * M, N)

        torch.clamp(self.w, 1e-6)
        cos_sim_matrix = cos_sim_matrix * self.w + self.b
        
        label = torch.from_numpy(np.asarray(range(N))).cuda()  # (N, )
        labels = torch.repeat_interleave(label, repeats=M, dim=0).cuda()  # (N * M)

        loss = self.criterion(cos_sim_matrix, labels)
        return loss
```
